### DISCUSSION TITLE
JsonNullable and Instancio

### DISCUSSION THREAD

**Discussion by nazariy-codephusion** [instancio/instancio#593] [Q&A] [answered]
Hello,

we are using JsonNullable from https://github.com/OpenAPITools/jackson-databind-nullable.
our dto mapping is like 

```
public SomeDto {
@Date
private JsonNullable<String> startDate = JsonNullable.undefined();
}
```

without using JsonNullable we can achieve custom date generation by executing:
```

instancioApi.generate(Select.fields().ofType(String.class).annotated(Date.class),
          gen -> gen.temporal().localDate().asString(DateTimeFormatter.ISO_LOCAL_DATE::format));
```

Is it possible somehow to have it wrapped in JsonNullable and have correct date as string inside ?

Thanks.

**Comment by armandino** [comment, marked as answer]
@nazar-phokus there are a couple of options to achieve this.

- **Option 1:** if there are only a couple of specific fields, one option is to use [`within(Scope)`](https://www.instancio.org/user-guide/#selector-scopes):

```java
SomeDto result = Instancio.of(SomeDto.class)
    .generate(allStrings().within(field(SomeDto::getStartDate).toScope()),
            gen -> gen.temporal().localDate().as(DateTimeFormatter.ISO_LOCAL_DATE::format))
    .create();
```

Assuming the DTO has a getter, this targets strings within `field(SomeDto::getStartDate)`. Btw, not sure which version of Instancio you're using but `asString()` has been deprecated; you can replace it with `as()` instead.

- **Option 2:** if there are many such fields and you need a more generic approach rather than targeting specific fields, then you could define a `Supplier` for all `JsonNullable` fields annotated with `@Date`:

```java
Supplier<JsonNullable<String>> supplier = () -> JsonNullable.of(
    DateTimeFormatter.ISO_LOCAL_DATE.format(LocalDate.now()));

SomeDto result = Instancio.of(SomeDto.class)
    .supply(fields().ofType(JsonNullable.class).annotated(Date.class), supplier)
    .create();
```

If you need random dates, you could define a `Generator` instead of a `Supplier`, with everything else remaining the same.

Hope this helps!

**Reply by nazariy-codephusion** [reply]
Thanks for the quick feedback
